### ISSUE TITLE
Parallel Queries and Temporary Query Result Error

### ISSUE THREAD

**Issue by ovflowd** [gatsbyjs/gatsby#36599]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Our Gatsby application suffers from crashes (100% 10/10) when running `gatsby build` due to "Parallel Queries" crashing without saying which query is crashing.

And during development runtimes, randomly, when opening different pages, it just crashes most of the time with `Error: Couldn't find temp query result for "/en/blog/releases/v16.6.2".` (Or similar, with the difference of other page titles).

Note that by removing the newly added `content/blog/releases` the crash from `gatsby build` disappears, but the random temp query result issue persists. I have no idea which file is causing it to crash. All those files abide by MDXv2 standards. (Even tho the App uses MDXv1)

### Reproduction Link

https://github.com/ovflowd/nodejs.dev/tree/test/gatsby-crashes

### Steps to Reproduce

1. Clone, checkout
2. Run `gatsby build`
3. Crashes

OR

1. Clone, checkout
2. Run `gatsby develop`
3. Navigate to different blog posts
4. Crashes

### Expected Result

1. To not crash?

### Actual Result

1. It crashes.

### Environment

```shell
  System:
    OS: macOS 12.5.1
    CPU: (10) arm64 Apple M1 Max
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.9.0 - /var/folders/kr/dtpb07wj0cl0vp56klz_xc440000gq/T/yarn--1663025110592-0.9226765595115247/node
    Yarn: 1.22.19 - /var/folders/kr/dtpb07wj0cl0vp56klz_xc440000gq/T/yarn--1663025110592-0.9226765595115247/yarn
    npm: 8.19.1 - /opt/homebrew/bin/npm
  Browsers:
    Chrome: 105.0.5195.102
    Firefox: 103.0.2
    Safari: 15.6.1
  npmPackages:
    gatsby: ^4.22.1 => 4.22.1 
    gatsby-link: ^4.22.0 => 4.22.0 
    gatsby-plugin-canonical-urls: ^4.22.0 => 4.22.0 
    gatsby-plugin-catch-links: ^4.22.0 => 4.22.0 
    gatsby-plugin-manifest: ^4.22.0 => 4.22.0 
    gatsby-plugin-mdx: ^3.20.0 => 3.20.0 
    gatsby-plugin-meta-redirect: ^1.1.1 => 1.1.1 
    gatsby-plugin-offline-next: ^5.2.3 => 5.2.3 
    gatsby-plugin-sass: ^5.22.0 => 5.22.0 
    gatsby-plugin-sitemap: ^5.22.0 => 5.22.0 
    gatsby-plugin-svgr: ^3.0.0-beta.0 => 3.0.0-beta.0 
    gatsby-plugin-typescript: ^4.22.0 => 4.22.0 
    gatsby-plugin-webfonts: ^2.2.2 => 2.2.2 
    gatsby-remark-autolink-headers: ^5.22.0 => 5.22.0 
    gatsby-remark-prismjs: ^6.22.0 => 6.22.0 
    gatsby-source-filesystem: ^4.22.0 => 4.22.0 
    gatsby-theme-i18n: ^3.0.0 => 3.0.0 
    gatsby-transformer-yaml: ^4.22.0 => 4.22.0 

```

### Config Flags

FAST_DEV: true

**Comment by ovflowd** [comment]
Hey @LekoArts, sorry for pinging you, but this is causing disruption for the Node.js org. Could you maybe give an 👀 here on your earlier convenience? Thanks 🙇

**Comment by ovflowd** [comment]
Will try to do, ty!

**Comment by ovflowd** [comment]
Note.: It only happens on build, when just running dev server, all pages work nicely.

**Comment by graysonhicks** [comment]
Hi @ovflowd !  I gave this a run locally and believe the MDX v2 syntax in the `content/blog/releases` folder is causing the issue.  I ran the posts through an MDX validator and found the following syntax errors:
- `<br>` tag wasn't closed (replaced in all files with `<br />`
-  the use of `<autolink>` tags, [which aren't supported in v1](https://github.com/mdx-js/mdx/issues/1049) is found [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2021-12-17-v17.3.0.md?plain=1#L11) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-03-18-v16.14.2.md?plain=1#L12) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-03-18-v17.7.2.md?plain=1#L12) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-04-19-v18.0.0.md?plain=1#L69) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-04-19-v18.0.0.md?plain=1#L98) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-04-19-v18.0.0.md?plain=1#L106) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-05-17-v18.2.0.md?plain=1#L12) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-06-02-v17.9.1.md?plain=1#L14) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-07-13-v18.6.0.md?plain=1#L19)
- several <autolinks> to bots in [this one](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-06-01-v16.15.1.md)
- unescaped `(` [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2021-12-17-v17.3.0.md?plain=1#L34) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-02-08-v16.14.0.md?plain=1#L92)
- unescaped `{` [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-02-10-v17.5.0.md?plain=1#L211) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-02-10-v17.5.0.md?plain=1#L213) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-03-16-v16.14.1.md?plain=1#L166) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-03-16-v16.14.1.md?plain=1#L168) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-07-13-v18.6.0.md?plain=1#L190) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-08-16-v16.17.0.md?plain=1#L698)
- remove escape and add closing /> for `td` [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-02-23-v17.6.0.md?plain=1#L59)
- replace escaped angle brackets with `&lt;` and `&gt` for Typescript type (or wrap in string completely) [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-03-22-v17.8.0.md?plain=1#L67) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-04-27-v16.15.0.md?plain=1#L237) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-08-16-v16.17.0.md?plain=1#L237) and [here](https://github.com/ovflowd/nodejs.dev/blob/test/gatsby-crashes/content/blog/releases/2022-07-13-v18.6.0.md?plain=1#L62)

After fixing those locally, my build succeeds.  Long-term, you'll need to ensure that your MD is outputted as compliant with MDX V1, *or*, upgrade your Gatsby MDX setup to V2 (but that has backwards breaking changes for the MD files not in the `release` dir).  I've got the fixed files in my fork [here](https://github.com/graysonhicks/nodejs.dev/tree/test/gatsby-crashes/content/blog/releases).

**Comment by ovflowd** [comment]
Hey, @graysonhicks thank you so much for actually spending time on this and doing such a deep analysis, this is spectacular! Regarding making all the Markdown files MDX v1 compliant. At least on the `main` branch, (not sure if in the linked branch too) we already use MDX v1 validation [here](https://github.com/nodejs/nodejs.dev/blob/main/package.json#L84) that comes from [here](https://github.com/nodejs/nodejs.dev/blob/main/.eslintrc.js#L59) (Ignore the comment there it is outdated).

So in theory all Markdown files should already be MDX v1 compliant. Is there any way we could automatically fix these issues? As we're planning into migrating all the blog posts from nodejs.org to there, this sounds like a massive work 😅

Not to mention, more API docs are going to be moved, as the script right now only generates v18 docs, but if I modify the script to also add the v16 docs, the build also crashes. It is kinda tricky to have MDX just right-away crashing when it finds non-compliant stuff. Other Markdown parsers don't do that... I do understand we should be compliant, and if there's an easier way of doing so, I'm down, and I realise now this is not a Gatsby specific issue but MDX one...

I'm definitely open for tips/advice :)

**Comment by graysonhicks** [comment]
No problem @ovflowd .  I only tested on the `test/gatsby-crash` branch.  How are the MD files in `releases` being generated?  That's likely where a change needs to happen, maybe something changed to start exporting the files in MDX v2 format.

**Comment by ovflowd** [comment]
> How are the MD files in releases being generated?

They're not being generated. Basically "copy/pasted" from [here](https://github.com/nodejs/nodejs.org/tree/main/locale/en/blog). Then we run a [script](https://github.com/nodejs/nodejs.dev/blob/main/scripts/transformNodeBlogPosts.js) that modifies the frontmatter to match the ones from our blog. Also, we run our lint command to check they respect all rules; otherwise, we run `--fix`, but in my experience, it doesn't necessarily fix all things.

What is generated are our API docs through [this script](https://github.com/nodejs/nodejs.dev/blob/main/scripts/syncApiDocs.js). And those are heavily modified through a few regexe's to be MDX compliant, or at least as much as possible. The lint command also doesn't give any error to them.

**Comment by ovflowd** [comment]
> maybe something changed to start exporting the files in MDX v2 format.

They're not being exported on MDXv2. We never attempted to go through the MDXv2 format. (I tried, and checked with linting to MDXv2 how much change would be needed, and ehhh, 1k+ errors straight of the bat, the TL;DR is here: https://github.com/gatsbyjs/gatsby/discussions/36519#discussioncomment-3618979)

**Comment by ovflowd** [comment]
@graysonhicks is there any way I can debug which file threw a parallel query error?

**Comment by ovflowd** [comment]
I'm getting the crash now in another branch, where the only thing I did was to rename some files.

**Comment by graysonhicks** [comment]
@ovflowd Thanks for pointing to that larger discussion, I think I found and fixed a lot of the errors you had already seen 😅 .  Interestingly, you note in that thread that you know there is syntax breaking MDXv2, but it appears to be breaking in MDXv1 based on my testing today as well.  Ideally you would have all the content be fixed, as Gatsby expects valid MD(X) files.  I agree there could be better error handling on the particular file.  I ran them through a compiler [here](https://mdxjs.com/playground/).  Did anything change with how `release` MD was written?  You may want to look at an MDX linter since you are using gatsby-plugin-mdx on MD files and maybe that's why I saw some errors in the compiler that you weren't seeing with your linter?

**Comment by graysonhicks** [comment]
Confirmed several of the issues I found were syntax errors introduced in MDXv2, so I'm not sure why those would be failing on MDXv1:

https://mdxjs.com/docs/troubleshooting-mdx/#unexpected-end-of-file-in-expression-expected-a-corresponding-closing-brace-for-

**Comment by ovflowd** [comment]
> . Did anything change with how release MD was written?

Nope. As I mentioned, the `mdx` parser for eslint gives 0 errors about any of those files... (Or if they did, I think I fixed them). But it is honestly a bummer that Gatsby itself crashes (Worker exits unsuccessfully) without even pointing out the error or where. It is tough to debug like that. Because this means any new MDX file could potentially break the whole App.

And sadly, we cannot migrate to MDXv2 as stated in the thread, or at least gatsby-plugin-mdx@4 because it has to break changes that are just entirely incompatible with, for example, i18n.

**Comment by ovflowd** [comment]
*Note.:* The way how we do i18n right now is every page makes the query of the same MDX document in the selected language and the current language. If a file with the current language is not found, we fall back to the default language file. With the MDX files being assigned directly on `gatsby-node` I find it really hard to actually create the pages in all languages. It would require me to make even more patches on the source of `gatsby-theme-i18n`. Or just abandon that plugin at all and make our own.

**Comment by LekoArts** [comment]
If I understand it correctly then, we can close this issue and move further discussion to "Discussions" as the root of the issue was invalid MDX? Our workers should throw more descriptive errors, if we can have a reproduction of a swallowed error (showing the above error, not the original one) we can discuss that in a new issue.

**Comment by ovflowd** [comment]
>  as the root of the issue was invalid MDX

I don't think so, again, MDX parser and linter doesn't complain about the contents I added.

> Our workers should throw more descriptive errors,

But they don't. They gave nothing, and I'm not exaggerating. Here's an example:

```
success Building HTML renderer - 19.096s
success Execute page configs - 0.044s
success Caching Webpack compilations - 0.001s

 ERROR #85928 

An error occurred during parallel query running.
Go here for troubleshooting tips: https://gatsby.dev/pqr-feedback



  Error: Worker exited before finishing task
  
  - index.js:117 ChildProcess.<anonymous>
    [nodejs.dev]/[gatsby-worker]/dist/index.js:117:45
  
  - node:events:513 ChildProcess.emit
    node:events:513:28
  
  - child_process:291 Process.ChildProcess._handle.onexit
    node:internal/child_process:291:12
  

not finished run queries in workers - 0.177s
```

> please note though that we're a small team and there's no guarantee to always get a promptly answer as such we partially rely on community members helping each other.

I understand that, and I appreciate the help. I'm just honestly out of options 😅 (As I honestly have no idea what to do with this error)

**Comment by graysonhicks** [comment]
So I was still unable to find a local linting solution that showed the same errors I was able to find by running through the MDX compiler.  *But*, I reintroduced all of the syntax errors and bumped these dependencies:

```
    "eslint-mdx": "^2.0.0"
    "eslint-plugin-mdx": "^2.0.0"
```

It then builds successfully for me. Can you try that?

**Comment by graysonhicks** [comment]
The `test/gatsby-crashes` branch builds for me successfully with that change (or with the syntax changes).

**Comment by ovflowd** [comment]
How could an eslint-plugin change make the builds pass? Am I missing something? 😅

**Comment by graysonhicks** [comment]
I don't know, I'm just letting you know what can unblock you.

**Comment by ovflowd** [comment]
Gotcha, sorry for the late reply. I'm going to do what you did then, and see what happens. 🙏

**Comment by ovflowd** [comment]
Let's close this for now, as there are no actual paths for fixing any of these issues.
